src/sentry/static/sentry/app/views/dashboardsV2/dashboard.tsx

```diff
@@ -146,294 +104,86 @@ class Dashboard extends React.Component<Props, State> {
-      draggingOverWidget: undefined,
-      ghostTopOffset: 0,
-      ghostLeftOffset: 0,
-      widgets: undefined,
-    });
-  };
-
   renderWidget(widget: Widget, index: number) {
     const {isEditing} = this.props;
 
+    const key = generateWidgetId(widget, index);
+    const dragId = key;
+
     return (
-      <LazyLoad key={`${widget.id ?? index}`} once height={240} offset={100}>
-        <WidgetWrapper data-component="widget-wrapper" displayType={widget.displayType}>
-          <WidgetCard
-            widget={widget}
-            isEditing={isEditing}
-            onDelete={this.handleDeleteWidget(index)}
-            onEdit={this.handleEditWidget(widget, index)}
-            isDragging={this.state.isDragging && this.state.draggingTargetIndex === index}
-            hideToolbar={this.state.isDragging}
-            startWidgetDrag={this.startWidgetDrag(index)}
-          />
-        </WidgetWrapper>
+      <LazyLoad key={key} once height={240} offset={100}>
+        <SortableWidget
+          widget={widget}
+          dragId={dragId}
+          isEditing={isEditing}
+          isDragging={this.state.isDragging}
+          onDelete={this.handleDeleteWidget(index)}
+          onEdit={this.handleEditWidget(widget, index)}
+        />
       </LazyLoad>
     );
   }
 
-  renderGhost = () => {
-    if (!this.portal) {
-      return null;
-    }
-    const {ghostLeftOffset, ghostTopOffset} = this.state;
-
-    const top = typeof this.state.top === 'number' ? this.state.top - ghostTopOffset : 0;
-    const left =
-      typeof this.state.left === 'number' ? this.state.left - ghostLeftOffset : 0;
-
-    const ghost = (
-      <WidgetGhost
-        id="ghost"
-        ref={this.dragGhostRef}
-        style={{top: `${top}px`, left: `${left}px`}}
-      />
-    );
-
-    return ReactDOM.createPortal(ghost, this.portal);
-  };
-
   render() {
     const {isEditing, dashboard} = this.props;
 
-    const widgets =
-      this.state.isDragging && this.state.widgets
-        ? this.state.widgets
-        : dashboard.widgets;
-
     return (
-      <WidgetContainer>
-        {widgets.map((widget, i) => this.renderWidget(widget, i))}
-        {isEditing && (
-          <WidgetWrapper key="add" displayType="big_number">
-            <AddWidgetWrapper
-              key="add"
-              data-test-id="widget-add"
-              onClick={this.handleStartAdd}
-            >
-              <IconAdd size="lg" isCircled color="inactive" />
-            </AddWidgetWrapper>
-          </WidgetWrapper>
-        )}
-        {this.renderGhost()}
-      </WidgetContainer>
+      <DndContext
+        onDragStart={({active}) => {
+          if (!active) {
+            return;
+          }
+
+          this.setState({
+            isDragging: true,
+            activeDragId: active.id,
+          });
+        }}
+        onDragEnd={({over}) => {
+          const {activeDragId} = this.state;
+          const items = this.getWidgetIds();
```

Comment:
should we move this out of this function and reuse it here https://github.com/getsentry/sentry/blob/f90176afd7585cf1cf859dbaa89580ba37160edd/src/sentry/static/sentry/app/views/dashboardsV2/dashboard.tsx#L181 ?